mavros_extras/src/plugins/esc_telemetry.cpp

```diff
@@ -21,6 +25,90 @@ namespace mavros
 {
 namespace extra_plugins
 {
+// for all the dignostics parameters
+struct Limits {
+	bool diag_enabled;
+	// Temperature checks
+	float temp_min_nok;
+	float temp_min_ok;
+	float temp_max_nok;
+	float temp_max_ok;
+	// Voltage checks
+	float volt_min_nok;
+	float volt_min_ok;
+	float volt_max_nok;
+	float volt_max_ok;
+	// Current checks
+	float curr_min_nok;
+	float curr_min_ok;
+	float curr_max_nok;
+	float curr_max_ok;
+	// RPM checks
+	int rpm_min_nok;
+	int rpm_min_ok;
+	int rpm_max_nok;
+	int rpm_max_ok;
+	// Count checks, should be uint16_t but ROS only allows int parameters
+	int count_min_delta;
+
+	explicit Limits(ros::NodeHandle &pnh) {
+		if (!pnh.getParam("enabled", this->diag_enabled)) {
+			ROS_FATAL_STREAM("Could not retrieve 'enabled' parameter from parameter server");
+			this->diag_enabled = false;
+			return;
+		}
+		if (this->diag_enabled)
+		{
+			// [[[cog:
+			// fields = [
+			//   ("temp", (1.0, 0.0), (85.0, 90.0), ),
+			//   ("volt", (15.0, 14.0), (16.0, 17.0), ),
+			//   ("curr", (5.0, 4.0), (8.0, 10.0), ),
+			//   ("rpm", (3000, 2000), (9000, 12000), )
+			// ]
+			//
+			// for field, min_, max_ = range fields:
+			//    for fm, lim = zip(("min", "max"), (min_, max_)):
+			//        for fn, l = zip(("ok", "nok"), lim):
+			//            cog.outl(f"""pnh.param("{field}_{fm}/{fn}", this->{field}_{fm}_{fn}, {l});""")
+			// ]]]
+
+			// [[[end]]]
+/**/
+			pnh.param("temp_min/nok", this->temp_min_nok, 0.0f);
+			pnh.param("temp_min/ok", this->temp_min_ok, 1.0f);
+			pnh.param("volt_min/nok", this->volt_min_nok, 14.0f);
+			pnh.param("volt_min/ok", this->volt_min_ok, 15.0f);
+			pnh.param("curr_min/nok", this->curr_min_nok, 4.0f);
+			pnh.param("curr_min/ok", this->curr_min_ok, 5.0f);
+			pnh.param("rpm_min/nok", this->rpm_min_nok, 2000);
+			pnh.param("rpm_min/ok", this->rpm_min_ok, 3000);
+			pnh.param("temp_max/nok", this->temp_max_nok, 90.0f);
+			pnh.param("temp_max/ok", this->temp_max_ok, 85.0f);
+			pnh.param("volt_max/nok", this->volt_max_nok, 17.0f);
+			pnh.param("volt_max/ok", this->volt_max_ok, 16.0f);
+			pnh.param("curr_max/nok", this->curr_max_nok, 10.0f);
+			pnh.param("curr_max/ok", this->curr_max_ok, 8.0f);
+			pnh.param("rpm_max/nok", this->rpm_max_nok, 12000);
+			pnh.param("rpm_max/ok", this->rpm_max_ok, 9000);
+/**/
+
+			pnh.param("count/min_delta", this->count_min_delta, 10);
+		}
+	}
+};
+
+class ESCDiag //: public diagnostic_updater::DiagnosticTask
+{
+public:
+	ESCDiag(const std::string& name, const Limits& lim_):
+		//diagnostic_updater::DiagnosticTask(name),
+		lim(lim_)
+	{}
+
+	const Limits& lim;
```

Comment:
Ah no, we can't really use const ref on temporary object. https://stackoverflow.com/questions/15513734/const-reference-as-class-member

Try to add limits object to the plugin class and pass it to diags.